docs/features/component-versioning.md

```diff
@@ -0,0 +1,88 @@
+
+`[RoslynFeature("Generator.PostInitialization, LanguageVersion9")]`
+
+_(__Alternative__: this could directly be an array, rather than separated list?)_
+
+These features would be defined in the compiler, and passed to the component loader. The component loader would check that the requirements are met, and issue a warning when unable to meet them. The requirements would be accessible via API to enable the author to strongly reference the requirement (either an Enum or set of readonly properties).
+
+The exact requirement list is TBD, but would likely include language version and specific component features we introduce going forward. These features could be dynamically calculated by the compiler: for instance, whilst still in preview a feature might only be enabled if the selected language version is also preview.
+
+While we'd allow authors to key on language version, we wouldn't be changing any of the existing functionality in the compiler to use these requirements: it would purely be for making decisions about whether to load the component or not.
+
+Features would also be able to represent super-sets of other features, for instance `LanguageVersion9` would imply `LanguageVersion1..8`. Components without the attribute would always be unconditionally loaded, with the same behavior as today.
+
+### Selective disabling
+
+A second use case for this feature would be to selective _disable_ the loading of certain components.
+
+Consider the migration of an `ISourceGenerator` to `IIncrementalGenerator`: the author may wish to create a new `IIncrementalGenerator`, but still fall back to the original `ISourceGenerator` in older compilers. The above proposal allows the user to *enable* the incremental generator when supported, but conversely they also need a mechanism to *disable* the `ISourceGenerator`.
+
+The `[RoslynFeature]` attribute would support a second, optional parameter that would cause the component to **not** load when the given feature is present.
+
+`[RoslynFeature("", disableWhen: "IncrementalGenerator")]`
+
+This would allow the author to provide a pair of components that enable/disable depending on features, allowing a smoother upgrade path.
+
+### Disable Warning
+
+In the above case, the author would not want the compiler to warn about either generator not loading, because a fallback is in place in both cases. The attribute would have a third optional parameter that would suppress the production of a warning:
+
+`[RoslynFeature("", warnOnNotLoaded: false)]`
+
+## Alternative Approaches
+
+There are a couple of obvious alternatives to the above proposal:
+
+### Version numbers
+
+Instead of being feature based, we could instead allow the author to specify the Minimum (and/or maximum) roslyn version that is required for the compiler.
```

Comment:
We should clarify that this alternative approach has the advantage of not requiring any change for any current analyzer or source generator author. By simply updating the compiler to report a better error message on load failures, component versioning based on Roslyn versions is complete.